pm/queue.go

```diff
@@ -0,0 +1,148 @@
+// A recipient will have a ticketQueue per sender that it is actively receiving tickets from.
+// If a sender's max float is insufficient to cover the face value of a ticket it is added to the queue.
+// A ticket is pulled from the queue by the recipient when a sender has sufficient max float to cover
+// the next ticket in the queue
+//
+// Based off of: https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/queue.go
+type ticketQueue struct {
+	queue []*SignedTicket
+
+	// queueLen is an internal length counter that keeps track
+	// of the size of the queue. We maintain this counter instead
+	// of reading len(queue) in order to avoid locking the queue loop goroutine
+	queueLen int32
+
+	// in is a channel that receives queued tickets
+	in chan *SignedTicket
+
+	// availableMaxFloat is a channel that the queue uses
+	// to receive the latest available max float value
+	availableMaxFloat chan *big.Int
+
+	// redeemable is a channel that a recipient will receive
+	// redeemable tickets on as a sender's max float becomes
+	// sufficient to cover the face value of tickets
+	redeemable chan *SignedTicket
+
+	quit chan struct{}
+}
+
+func newTicketQueue() *ticketQueue {
+	return &ticketQueue{
+		in:                make(chan *SignedTicket),
+		availableMaxFloat: make(chan *big.Int),
+		redeemable:        make(chan *SignedTicket),
+		quit:              make(chan struct{}),
+	}
+}
+
+// Start initiates the main queue loop goroutine for processing tickets
+func (q *ticketQueue) Start() {
+	go q.startQueueLoop()
+}
+
+// Stop signals the ticketQueue to gracefully shutdown
+func (q *ticketQueue) Stop() {
+	close(q.quit)
+}
+
+// Add adds a ticket to the queue
+func (q *ticketQueue) Add(ticket *SignedTicket) {
+	select {
+	case q.in <- ticket:
+	case <-q.quit:
+		return
+	}
+
+	atomic.AddInt32(&q.queueLen, 1)
+}
+
+// SignalMaxFloat signals to the queue the latest max float for the sender
+// An external caller should call this method whenever a ticket redemption confirms
+// with the most up-to-date max float for the sender
+func (q *ticketQueue) SignalMaxFloat(amount *big.Int) {
+	// If the queue is empty do not signal a new max float
+	// value because there are no tickets to process
+	if q.Length() == 0 {
+		return
+	}
+
+	select {
+	case q.availableMaxFloat <- amount:
+	case <-q.quit:
+		return
+	}
+}
+
+// Redeemable returns a channel that a consumer can use to receive tickets that
+// should be redeemed
+func (q *ticketQueue) Redeemable() chan *SignedTicket {
+	return q.redeemable
+}
+
+// Length returns the current length of the queue
+func (q *ticketQueue) Length() int32 {
+	return atomic.LoadInt32(&q.queueLen)
+}
+
+func (q *ticketQueue) startQueueLoop() {
+	for {
+		select {
+		case ticket := <-q.in:
+			q.queue = append(q.queue, ticket)
+		case <-q.quit:
+			return
+		default:
+		}
+
+		if len(q.queue) == 0 {
+			continue
+		}
+
+		nextTicket := q.queue[0]
+
+		select {
+		case maxFloat := <-q.availableMaxFloat:
```

Comment:
Semantically, it feels a little strange reading this: we're waiting on a channel, but `availableMaxFloat` implies the value is immediately available (when it may not be). Would renaming the channel to something like `maxFloatUpdate` work?

Reply:
Fixed in 1d23297